File: ruby/simple-cipher/simple_cipher.rb
# frozen_string_literal: true

require 'pry'

class Cipher
  attr_reader :key

  ALPHABET = ('a'..'z').to_a
  A = 'a'.ord

  def initialize(key = nil)
    @key = key || generate_key

    raise ArgumentError unless valid?(@key)
  end

  def encode(text)
    text.chars.map.with_index do |char, index|
      ALPHABET.rotate(@key[index].ord - A)[char.ord - A]
    end.join
  end

  def decode(text)
    text.chars.map.with_index do |char, index|
      ALPHABET.rotate(A - @key[index].ord)[char.ord - A]
    end.join
  end

  private

  def valid?(key)
    /\A[a-z]+\z/ =~ key
  end

  def generate_key
    100.times.map { ALPHABET.sample }.join
  end
end

module Version
  VERSION = 1
end

File: python/simple-cipher/simple_cipher.py
from random import choice
from string import ascii_lowercase as alphabet
from itertools import cycle
import pdb


class Cipher(object):
    def __init__(self, key=None):
        self.key = key if key else self.generate_key()

    def encode(self, text):
        encoded = ''
        for c, k in zip(text, cycle(self.key)):
            encoded += alphabet[(alphabet.index(c) + alphabet.index(k)) % 26]
        return encoded

    def decode(self, text):
        decoded = ''
        for c, k in zip(text, cycle(self.key)):
            decoded += alphabet[(alphabet.index(c) - alphabet.index(k)) % 26]
        return decoded

    @staticmethod
    def generate_key():
        return ''.join(choice(alphabet) for _ in range(100))


class Version():
    version = 1
